packages/cli/src/databases/migrations/sqlite/1724951148974-AddApiKeysTable.ts

```diff
@@ -0,0 +1,74 @@
+import type { ApiKey } from '@/databases/entities/api-key';
+import type { MigrationContext } from '@/databases/types';
+import { generateNanoId } from '@/databases/utils/generators';
+
+export class AddApiKeysTable1724951148974 {
+	async up({ queryRunner, tablePrefix, runQuery }: MigrationContext) {
+		const tableName = `${tablePrefix}user_api_keys`;
+
+		// Create the table
+		await queryRunner.query(`
+			CREATE TABLE ${tableName} (
+				id VARCHAR(36) PRIMARY KEY NOT NULL,
+				"userId" VARCHAR NOT NULL,
+				"label" VARCHAR(255) NOT NULL,
+				"apiKey" VARCHAR NOT NULL,
+				"createdAt" DATETIME(3) NOT NULL DEFAULT (STRFTIME('%Y-%m-%d %H:%M:%f', 'NOW')),
+				"updatedAt" DATETIME(3) NOT NULL DEFAULT (STRFTIME('%Y-%m-%d %H:%M:%f', 'NOW')),
+				FOREIGN KEY ("userId") REFERENCES user(id) ON DELETE CASCADE,
+				UNIQUE ("userId", label),
+				UNIQUE("apiKey")
+			);
+		`);
+
+		const usersWithApiKeys = (await queryRunner.query(
+			`SELECT id, "apiKey" FROM ${tablePrefix}user WHERE "apiKey" IS NOT NULL`,
+		)) as Array<Partial<ApiKey>>;
+
+		// Move the apiKey from the users table to the new table
+		usersWithApiKeys.forEach(async (user: { id: string; apiKey: string }) => {
+			await runQuery(
+				`INSERT INTO ${tableName} ("id", "userId", "apiKey", "label") VALUES (:id, :userId, :apiKey, :label)`,
+				{
+					id: generateNanoId(),
+					userId: user.id,
+					apiKey: user.apiKey,
+					label: 'My API Key',
+				},
+			);
+		});
+
+		// Create temporary table to store the users dropping the api key column
+		await queryRunner.query(`
+				CREATE TABLE users_new (
+					id varchar PRIMARY KEY,
+					email VARCHAR(255) UNIQUE,
+					"firstName" VARCHAR(32),
+					"lastName" VARCHAR(32),
+					password VARCHAR,
+					"personalizationAnswers" TEXT,
+					"createdAt" DATETIME(3) NOT NULL DEFAULT (STRFTIME('%Y-%m-%d %H:%M:%f', 'NOW')),
+					"updatedAt" DATETIME(3) NOT NULL DEFAULT (STRFTIME('%Y-%m-%d %H:%M:%f', 'NOW')),
+					settings TEXT,
+					disabled BOOLEAN DEFAULT FALSE NOT NULL,
+					"mfaEnabled" BOOLEAN DEFAULT FALSE NOT NULL,
+					"mfaSecret" TEXT,
+					"mfaRecoveryCodes" TEXT,
+					role TEXT NOT NULL
+			);
+		`);
+
+		//  Copy the data from the original users table
+		await queryRunner.query(`
+			INSERT INTO users_new ("id", "email", "firstName", "lastName", "password", "personalizationAnswers", "createdAt", "updatedAt", "settings", "disabled", "mfaEnabled", "mfaSecret", "mfaRecoveryCodes", "role")
+			SELECT "id", "email", "firstName", "lastName", "password", "personalizationAnswers", "createdAt", "updatedAt", "settings", "disabled", "mfaEnabled", "mfaSecret", "mfaRecoveryCodes", "role"
+			FROM ${tablePrefix}user;
+		`);
+
+		// Drop table with apiKey column
+		await queryRunner.query(`DROP TABLE ${tablePrefix}user;`);
+
+		// Rename the temporary table to users
+		await queryRunner.query('ALTER TABLE users_new RENAME TO user;');
```

Comment:
Unfortunately this is not enough. We need to update every single table that has a foreign key to `user` table to point to the new table

Reply:
Since we are not using typeorm but doing this as direct sql, why don't we modify the existing table instead of copying the data to the new table?

Reply:
On second thought I'm not 100% certain how sqlite behaves when foreign key have been disabled 🤔 Does it still keep the foreign keys intact when we delete a table an rename another table as that table?

Reply:
after running the migration, I tried to add a record to `auth_identity`, which references the user's id in the user's table, with an unexciting user id and got: 

![CleanShot 2024-09-26 at 07 03 05@2x](https://github.com/user-attachments/assets/779bdb81-eb34-4df5-8f03-04b7fb943b27)

Reply:
Yes seems to be working correctly! Thank you for testing ❤️